src/approaches/grammar_search_invention_approach.py

```diff
@@ -236,16 +350,29 @@ def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
             yield (predicate, cost)
             classifier = _ForallClassifier(predicate)
             yield (Predicate(str(classifier), [], classifier), cost)
+            if predicate.arity >= 2:
+                for idx in range(predicate.arity):
+                    uff_classifier = _UnaryFreeForallClassifier(predicate, idx)
+                    uff_predicate = Predicate(str(uff_classifier),
+                                              [predicate.types[idx]],
+                                              uff_classifier)
+                    yield (uff_predicate, cost)
 
 
-def _create_grammar(grammar_name: str, dataset: Dataset) -> _PredicateGrammar:
+def _create_grammar(grammar_name: str, dataset: Dataset,
+                    given_predicates: Set[Predicate]) -> _PredicateGrammar:
     if grammar_name == "holding_dummy":
         return _HoldingDummyPredicateGrammar(dataset)
     if grammar_name == "single_feat_ineqs":
-        return _SingleFeatureInequalitiesPredicateGrammar(dataset)
+        sfi_grammar = _SingleFeatureInequalitiesPredicateGrammar(dataset)
+        return _NegationPredicateGrammarWrapper(sfi_grammar)
     if grammar_name == "forall_single_feat_ineqs":
-        base = _SingleFeatureInequalitiesPredicateGrammar(dataset)
-        return _ForallPredicateGrammarWrapper(dataset, base)
+        given_grammar = _GivenPredicateGrammar(given_predicates)
```

Comment:
What's the idea here behind proposing given predicates? Aren't they already going to be given to / used by the operator learning always?

Reply:
Okay now I'm just confused due to the SkipGrammar below going and removing the given_predicates again. This portion needs some comments, I think.